java/src/main/native/src/CuFileJni.cpp

```diff
@@ -163,104 +165,125 @@ class cufile_file {
   }
 
   /**
-   * @brief Write a device buffer to the file.
+   * @brief Write a device buffer to a file.
    *
+   * @param path Absolute path of the file to write to.
    * @param buffer The device buffer to write.
    * @param file_offset Starting offset from which to write the file.
    */
-  void write(cufile_buffer const &buffer, std::size_t file_offset) {
-    auto const status =
-        cuFileWrite(cufile_handle_, buffer.device_pointer(), buffer.size(), file_offset, 0);
+  static void write(char const *path, cufile_buffer const &buffer, std::size_t file_offset) {
+    do_write(path, buffer, file_offset);
+  }
 
-    if (status < 0) {
-      if (IS_CUFILE_ERR(status)) {
-        CUDF_FAIL("Failed to write buffer to file: " + cuFileGetErrorString(status));
-      } else {
-        CUDF_FAIL("Failed to write buffer to file: " + cuFileGetErrorString(errno));
-      }
-    }
+  /**
+   * @brief Append a device buffer to a file.
+   *
+   * @param path Absolute path of the file to append to.
+   * @param buffer The device buffer to append.
+   * @return The file offset from which the buffer was appended.
+   */
+  static std::size_t append(char const *path, cufile_buffer const &buffer) {
+    return do_write(path, buffer);
+  }
 
-    CUDF_EXPECTS(status == buffer.size(), "Size of bytes written is different from buffer size");
+  // Disable copy (and move) semantics.
+  cufile_file(cufile_file const &) = delete;
+  cufile_file &operator=(cufile_file const &) = delete;
+
+  /** @brief Destroy the file wrapper by de-registering the cuFile handle and closing the file. */
+  ~cufile_file() {
+    cuFileHandleDeregister(cufile_handle_);
+    close(file_descriptor_);
   }
 
+private:
   /**
-   * @brief Append a device buffer to the file.
+   * @brief Write a device buffer to a file.
    *
+   * @param path Absolute path of the file to write to.
    * @param buffer The device buffer to write.
+   * @param file_offset Starting offset from which to write the file. If set to max std::size_t,
+   * append to the file.
    * @return The file offset from which the buffer was appended.
    */
-  std::size_t append(cufile_buffer const &buffer) {
-    struct stat stat_buffer;
-    auto const status = fstat(file_descriptor_, &stat_buffer);
-    if (status < 0) {
-      CUDF_FAIL("Failed to get file status for appending: " + cuFileGetErrorString(errno));
+  static std::size_t do_write(char const *path, cufile_buffer const &buffer,
+                              std::size_t file_offset = std::numeric_limits<std::size_t>::max()) {
+    auto file_descriptor = open(path, O_CREAT | O_WRONLY | O_DIRECT, S_IRUSR | S_IWUSR);
+    if (file_descriptor < 0) {
+      CUDF_FAIL("Failed to open file " + std::string(path) + ": " + cuFileGetErrorString(errno));
     }
 
-    auto const file_offset = static_cast<std::size_t>(stat_buffer.st_size);
-    write(buffer, file_offset);
-    return file_offset;
+    while (true) {
```

Comment:
Should there be a limit to the number of retries?  Wondering if there are cases where this could end up looping forever, and hanging can be worse than crashing in practice.